Цель: полномасштабно изучить новую библиотеку в python под названием pandas. с её помощью можно будет описать и изучить различные датасеты не только быстро, но и довольно удобно. Также не мешало бы и изучить некоторые тонкости языка при описании и работе с датасетами в теоритическом плане, после чего незамедлительно отработать все на практике. Нужно больше практики

In [24]:
import numpy as np
import pandas as pd

animes = pd.read_csv('anime.csv')
print(animes)
animes.describe()

       anime_id                                               name  \
0         32281                                     Kimi no Na wa.   
1          5114                   Fullmetal Alchemist: Brotherhood   
2         28977                                           Gintama°   
3          9253                                        Steins;Gate   
4          9969                                      Gintama&#039;   
...         ...                                                ...   
12289      9316       Toushindai My Lover: Minami tai Mecha-Minami   
12290      5543                                        Under World   
12291      5621                     Violence Gekiga David no Hoshi   
12292      6133  Violence Gekiga Shin David no Hoshi: Inma Dens...   
12293     26081                   Yasuji no Pornorama: Yacchimae!!   

                                                   genre   type episodes  \
0                   Drama, Romance, School, Supernatural  Movie        1   
1      

anime_id        rating       members
count  12294.000000  12064.000000  1.229400e+04
mean   14058.221653      6.473902  1.807134e+04
std    11455.294701      1.026746  5.482068e+04
min        1.000000      1.670000  5.000000e+00
25%     3484.250000      5.880000  2.250000e+02
50%    10260.500000      6.570000  1.550000e+03
75%    24794.500000      7.180000  9.437000e+03
max    34527.000000     10.000000  1.013917e+06

Данный датасет зранитт в себе записи огромного количества аниме, которое мог полностю посмотреть только истиный отаку. Первая колонка хранит в себе лишь номер анимешки, второй - оригинальное имя придуманое японскими/корейскими/китайскими гениями индустрии развлечений (самой разной). Долгожданный третий столбец содержит перечень основных жанров анимешки, по которым профессиональные отаку могут уже судить и предсказывать неколторые сюжетные моменты, например хентай. Столбец type говорит нам нам о формате выхода хронометража (так, TV - эо формат сериала, где серии выходят на тв с некоторым промежутко по времени; oca - 1-2-х серийные дополнения к сюжету основной истории, развивающие сюжет или фансервис; rating - показывает рейтинг аниме (сюда бесполезно смотреть, ибо каждое аниме богоподобно, исключая хентай и гарем); episodes - показывает, сколько эпизодов содержит данный тайтл; members - показывает количество персонажей в аниме)

In [25]:
print(animes.nunique())

anime_id    12294
name        12292
genre        3264
type            6
episodes      187
rating        598
members      6706
dtype: int64


здесь мы можем лицезреть подстчёт количества уникальных значений в каждом столбце. Так, почему-то у как минимум 2-ух аниме одинаковые названия, что не может не огорчать. остальное можно не смотреть, поскольку и жанры и количесвто персонажей и кол-во эпизодов и другая информация об аниме в данном датасете моджет повторяться, это же не сюжет (который иногда тоже повторяется).

In [26]:
print(animes['type'].unique())


['Movie' 'TV' 'OVA' 'Special' 'Music' 'ONA' nan]


Здесь же мы можем тщательно изучить все существующие форамты аниме. Сркди них: сериал, фильм, односерийник, веб-аниме, музыкальное аниме.

In [27]:
print(animes.sort_values(by='rating'))

       anime_id                                               name  \
9991      34476                 Platonic Chain: Ansatsu Jikkouchuu   
12258     20007                      Hi Gekiga Ukiyoe Senya Ichiya   
8058       3287                       Tenkuu Danzai Skelter+Heaven   
8057      13405                                 Utsu Musume Sayuri   
8056        413                                    Hametsu no Mars   
...         ...                                                ...   
12274     34492  Nuki Doki! Tenshi to Akuma no Sakusei Battle -...   
12279     34491                       Sagurare Otome The Animation   
12280     34312                                       Saimin Class   
12282     34388                    Shikkoku no Shaga The Animation   
12285     34399                                   Taimanin Asagi 3   

                              genre     type episodes  rating  members  
9991          Sci-Fi, Slice of Life  Special        1    1.67       51  
12258        

В данном отрывке кода располагается вывод анимешек по возрастанию их рейтинга средии пользователей. Это полезно, если вы доверяете мнению большинства и готовы возможно потерчять настоящий шедевр, неоцененный большинством

In [28]:
print(animes.shape)

(12294, 7)


показывет, сколько стольбцов (7) и сколько строк божественного контента мы можем найти в данном датасете (12294)

In [29]:
print(animes.columns)
headers = np.array(animes.columns)
definitions = np.array(['Обозначает лишь номер анимешки', 'оригинальное имя придуманое японскими/корейскими/китайскими непризнанными гениями', 'перечень основных жанров', 'тип аниме', 'количество эпизодов', 'оющее количество персонажей'])

for i in range(1, definitions.size):
    print(f'{headers[i]} - {definitions[i-1]}')

Index(['anime_id', 'name', 'genre', 'type', 'episodes', 'rating', 'members'], dtype='object')
name - Обозначает лишь номер анимешки
genre - оригинальное имя придуманое японскими/корейскими/китайскими непризнанными гениями
type - перечень основных жанров
episodes - тип аниме
rating - количество эпизодов


Таким образом мы просветили незнающих в каноны данного датасета и показаи ему истиное значение каждого столбца

In [30]:
print(animes['name'].unique())

['Kimi no Na wa.' 'Fullmetal Alchemist: Brotherhood' 'Gintama°' ...
 'Violence Gekiga David no Hoshi'
 'Violence Gekiga Shin David no Hoshi: Inma Densetsu'
 'Yasuji no Pornorama: Yacchimae!!']


Естественно, мы получим абсолютно все значения из датасета, поскольку каждый тайтл по природе уникеален и неповторим так же, как и его название. Ни один уважающий себя китайски/японский/корейский творец не породит копипаст и не украдет ничего у другого творца (кроме сюжета).

In [31]:
print(animes['genre'].unique())

['Drama, Romance, School, Supernatural'
 'Action, Adventure, Drama, Fantasy, Magic, Military, Shounen'
 'Action, Comedy, Historical, Parody, Samurai, Sci-Fi, Shounen' ...
 'Hentai, Sports' 'Drama, Romance, School, Yuri' 'Hentai, Slice of Life']


Мы смертные получим все уникальные комплектации жанров, которые могут быть у анимешки. Тут, к сожалению будут не все записи, поскольку аниме - это конечно искусство, но жанры могут повторяться.

In [32]:
print(animes[animes['genre'] == 'Fantasy'])

       anime_id                                           name    genre  \
415        7711                        Karigurashi no Arrietty  Fantasy   
1283       5998  Hottarake no Shima: Haruka to Mahou no Kagami  Fantasy   
1516      24459             Tales of Zestiria: Doushi no Yoake  Fantasy   
1526       2672   Doraemon Movie 05: Nobita no Makai Daibouken  Fantasy   
1962       2675     Doraemon Movie 17: Nobita to Ginga Express  Fantasy   
...         ...                                            ...      ...   
10885     12759                   Zoku Zoku Mura no Obaketachi  Fantasy   
10930     31629                 Granblue Fantasy The Animation  Fantasy   
10965     34346                     Dam Keeper no Nakama-tachi  Fantasy   
10967     34497  Death March kara Hajimaru Isekai Kyousoukyoku  Fantasy   
11093     21305                                 Trick or Alice  Fantasy   

          type episodes  rating  members  
415      Movie        1    8.10    97910  
1283     Movi

По совершении этой опреации мы на выход получим количество аниме относящихся к лучшему жанру на свете - фентези. При просмотре данного жанра можно с головой окнуться в мир ,полный тайн, загадок, магии и самых разных фантастических существ, воплотить самые наивные детские мечты. Возвращаясь к коду, операция сработала, ну и отлично, главное больше её не трогать.

In [33]:
animes.drop(columns=['anime_id'])

name  \
0                                         Kimi no Na wa.   
1                       Fullmetal Alchemist: Brotherhood   
2                                               Gintama°   
3                                            Steins;Gate   
4                                          Gintama&#039;   
...                                                  ...   
12289       Toushindai My Lover: Minami tai Mecha-Minami   
12290                                        Under World   
12291                     Violence Gekiga David no Hoshi   
12292  Violence Gekiga Shin David no Hoshi: Inma Dens...   
12293                   Yasuji no Pornorama: Yacchimae!!   

                                                   genre   type episodes  \
0                   Drama, Romance, School, Supernatural  Movie        1   
1      Action, Adventure, Drama, Fantasy, Magic, Mili...     TV       64   
2      Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
3                                       Sci-Fi, Thriller     TV       24   
4      Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
...                                                  ...    ...      ...   
12289                                             Hentai    OVA        1   
12290                                             Hentai    OVA        1   
12291                                             Hentai    OVA        4   
12292                                             Hentai    OVA        1   
12293                                             Hentai  Movie        1   

       rating  members  
0        9.37   200630  
1        9.26   793665  
2        9.25   114262  
3        9.17   673572  
4        9.16   151266  
...       ...      ...  
12289    4.15      211  
12290    4.28      183  
12291    4.88      219  
12292    4.98      175  
12293    5.46      142  

[12294 rows x 6 columns]

как бы ни было больно моей душе, но по заданию необходимо удпалить какой-либо столбец. Однако такая проблема: этот датасет посвящен аниме, а значит в нём не может по определенитю быть ничего лишнего. Но, буквально отрывая частичку себя с этим  проклятым столбцом, мы удаляем столбец с id фнимешки. Мне не сделжать слёз от такой утраты, они лъются рекой и их ничем не остановить. Это вынудило мнея отойти от работы,  выйти на балкон и под пост-панк просто смотреть на непримечательные повторяющиеся и затянутые легкой дымкой старые виды домов, ставшие моими большими друзьями. В такие моменты, когда ты просто устал от всей той горы проблем, навалившейся на твои плечи, хочется просто остаться на этом балконе и безудержно смотреть в даль, в опять застывший пейзаж в окне.

In [34]:
animes.replace({'genre': 'Hentai'}, "Heretic's sin")
animes.replace ({'genre': 'Fantasy'}, 'deep dark fantasy')
animes.replace({'type':'TV'}, 'TV-serial')

anime_id                                               name  \
0         32281                                     Kimi no Na wa.   
1          5114                   Fullmetal Alchemist: Brotherhood   
2         28977                                           Gintama°   
3          9253                                        Steins;Gate   
4          9969                                      Gintama&#039;   
...         ...                                                ...   
12289      9316       Toushindai My Lover: Minami tai Mecha-Minami   
12290      5543                                        Under World   
12291      5621                     Violence Gekiga David no Hoshi   
12292      6133  Violence Gekiga Shin David no Hoshi: Inma Dens...   
12293     26081                   Yasuji no Pornorama: Yacchimae!!   

                                                   genre       type episodes  \
0                   Drama, Romance, School, Supernatural      Movie        1   
1      Action, Adventure, Drama, Fantasy, Magic, Mili...  TV-serial       64   
2      Action, Comedy, Historical, Parody, Samurai, S...  TV-serial       51   
3                                       Sci-Fi, Thriller  TV-serial       24   
4      Action, Comedy, Historical, Parody, Samurai, S...  TV-serial       51   
...                                                  ...        ...      ...   
12289                                             Hentai        OVA        1   
12290                                             Hentai        OVA        1   
12291                                             Hentai        OVA        4   
12292                                             Hentai        OVA        1   
12293                                             Hentai      Movie        1   

       rating  members  
0        9.37   200630  
1        9.26   793665  
2        9.25   114262  
3        9.17   673572  
4        9.16   151266  
...       ...      ...  
12289    4.15      211  
12290    4.28      183  
12291    4.88      219  
12292    4.98      175  
12293    5.46      142  

[12294 rows x 7 columns]

Написанный ранее код меняет неекоторые жанры на более подходящие. Так, например, hentai меняется на перевод: грех еретика; фентези меняется на глубокое темное фентези, а тпи TV поменялся на tv-serial

In [35]:
animes.drop_duplicates()

anime_id                                               name  \
0         32281                                     Kimi no Na wa.   
1          5114                   Fullmetal Alchemist: Brotherhood   
2         28977                                           Gintama°   
3          9253                                        Steins;Gate   
4          9969                                      Gintama&#039;   
...         ...                                                ...   
12289      9316       Toushindai My Lover: Minami tai Mecha-Minami   
12290      5543                                        Under World   
12291      5621                     Violence Gekiga David no Hoshi   
12292      6133  Violence Gekiga Shin David no Hoshi: Inma Dens...   
12293     26081                   Yasuji no Pornorama: Yacchimae!!   

                                                   genre   type episodes  \
0                   Drama, Romance, School, Supernatural  Movie        1   
1      Action, Adventure, Drama, Fantasy, Magic, Mili...     TV       64   
2      Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
3                                       Sci-Fi, Thriller     TV       24   
4      Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
...                                                  ...    ...      ...   
12289                                             Hentai    OVA        1   
12290                                             Hentai    OVA        1   
12291                                             Hentai    OVA        4   
12292                                             Hentai    OVA        1   
12293                                             Hentai  Movie        1   

       rating  members  
0        9.37   200630  
1        9.26   793665  
2        9.25   114262  
3        9.17   673572  
4        9.16   151266  
...       ...      ...  
12289    4.15      211  
12290    4.28      183  
12291    4.88      219  
12292    4.98      175  
12293    5.46      142  

[12294 rows x 7 columns]

удалит все повторяющиеся записи, но, как и было сказано ранее: повторяшек там никогда не будет (как минимум в названии)

In [36]:
animes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12294 entries, 0 to 12293
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   anime_id  12294 non-null  int64  
 1   name      12294 non-null  object 
 2   genre     12232 non-null  object 
 3   type      12269 non-null  object 
 4   episodes  12294 non-null  object 
 5   rating    12064 non-null  float64
 6   members   12294 non-null  int64  
dtypes: float64(1), int64(2), object(4)
memory usage: 672.5+ KB


покажет некоторую информацию о столбцах, например количесвто ненулевых значений, и тип данных, а также занимаемый объм в KB. не думаю, что можно сделать какие-лиьо полезные выводы, касательно данной информации, лишь что в некоторых записях отсутствует жанр (скорее всего от того, что жанр, идеально описывавший данный тайтл, просто ещё не придумали умные дяденьки из Китая/Японии/Кореи), а также что в некоторых записях отсутствует рейтинг (гениальность нельзя оценить простыми циферками) и тип.

In [37]:
animes.describe()

anime_id        rating       members
count  12294.000000  12064.000000  1.229400e+04
mean   14058.221653      6.473902  1.807134e+04
std    11455.294701      1.026746  5.482068e+04
min        1.000000      1.670000  5.000000e+00
25%     3484.250000      5.880000  2.250000e+02
50%    10260.500000      6.570000  1.550000e+03
75%    24794.500000      7.180000  9.437000e+03
max    34527.000000     10.000000  1.013917e+06

Можно сделать некоторые выводы касательно данных, полученных при помощи describe. Так, средний рейтинг по всем аниме равен 6.473902, по любому всё портит хентай, котрый является лишь ошибкой порозденной   по незнанию гениев с востока. Самый низкий рейтинг у анимешки - 1.67, а самый высокий, что понятно, - 10. Первая четверь аниме обладает средним рейтингом в 5.88, первая половина - 6.57, а первые 3/4 - 7.18; персонаажей же меного в любом случае.

In [38]:

print(animes.loc[[0,12,12000,11000]])

print(animes.loc[animes['name'] == 'Rakudai Kishi no Cavalry', ['name','genre']])

print(animes.loc[animes['genre'] == 'Fantasy', ['name', 'rating', 'type']].sort_values(by='rating'))

       anime_id              name  \
0         32281    Kimi no Na wa.   
12          918           Gintama   
12000      3393              Dark   
11000     23279  Gyakusatsu Kikan   

                                                   genre   type episodes  \
0                   Drama, Romance, School, Supernatural  Movie        1   
12     Action, Comedy, Historical, Parody, Samurai, S...     TV      201   
12000                                             Hentai    OVA        2   
11000                              Psychological, Sci-Fi  Movie        1   

       rating  members  
0        9.37   200630  
12       9.04   336376  
12000    5.75     1310  
11000     NaN    12529  
                         name                                    genre
900  Rakudai Kishi no Cavalry  Action, Ecchi, Fantasy, Romance, School
                                                name  rating     type
9244                        Kirin no Otenki Mama-san    3.53  Special
8542                      

покажет нам во- первых определенные записи под номерами 0, 12, 12000, 11000 ивыведет всю информацию о них, после выведет жанры определенного тайтkа, такого, как доблесть рыцаря-неудачника. таким образом была продемонстрирована возможность выбирать определенного рода записи или выбирать их по условиям. Наконец выводятся все запсис жанра фентези и сортируются по их рейтингу соответственно. Так, можно было выбрать люьой критерий.

In [39]:
animes.to_csv('NotAllAnimes.csv')

Наконец, сохранениев другой файл

Вывод: был продемонстрирован практически полностью базовый функционал библиотеки pandas языка программироования Python. Также была изучена и закреплена практикой возможность обозревания и описания датасета при помощи данного языка и данной библиотеки. Было выявлено множество новых возможностей, открывающий предо мной необъятный горизонт возможностей работы с таблицами и датасетами.